File: src/main/java/fit/wenchao/http_file_server/rest/FileController.java
package fit.wenchao.http_file_server.rest;

import fit.wenchao.http_file_server.ConfigFile;
import fit.wenchao.http_file_server.SystemProperty;
import fit.wenchao.http_file_server.constants.FileType;
import fit.wenchao.http_file_server.constants.RespCode;
import fit.wenchao.http_file_server.exception.BackendException;
import fit.wenchao.http_file_server.model.JsonResult;
import fit.wenchao.http_file_server.model.vo.FileInfo;
import fit.wenchao.http_file_server.model.vo.UploadFileInfo;
import fit.wenchao.http_file_server.utils.FilePathBuilder;
import fit.wenchao.http_file_server.utils.ResponseEntityUtils;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Null;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.nio.file.*;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static fit.wenchao.http_file_server.utils.StrUtils.filePathBuilder;
import static org.springframework.http.HttpHeaders.CONTENT_LENGTH;

@Validated
@RestController
public class FileController {

    public static void main(String[] args) {
        System.out.println();
    }

    @Autowired
    ConfigFile configFile;

    public String getRootPath() {
        String root = configFile.getProp("root");
        if (root == null || "".equals(root)) {
            root = System.getProperty("user.home");
        }

        return root;
    }

    @GetMapping("/file-list")
    public ResponseEntity<JsonResult> getFileList(@NotBlank String path) {

        String rootPath = getRootPath();
        if (rootPath == null) {
            throw new BackendException(rootPath, RespCode.ROOT_PATH_CONFIG_ERROR);
        }
        path = FilePathBuilder.ofPath()
                              .ct(rootPath)
                              .ct(path)
                              .build();

        File file = new File(path);

        // path is not valid
        if (!file.exists()) {
            throw new BackendException(path, RespCode.NO_FILE);
        }


        File[] files = file.listFiles();


        List<FileInfo> fileInfoList = new ArrayList<>();


        for (File everyFile : files) {
            String name = everyFile.getName();
            String absolutePath = everyFile.getAbsolutePath();
            long length = everyFile.length();
            boolean directory = everyFile.isDirectory();

            FileInfo fileInfo = FileInfo.builder()
                                        .name(name)
                                        .path(absolutePath)
                                        .length(length)
                                        .type(directory ?
                                                FileType.DIR.getCode() :
                                                FileType.FILE.getCode())
                                        .build();

            fileInfoList.add(fileInfo);
        }

        return ResponseEntityUtils.ok(JsonResult.ok(fileInfoList));
    }

    @GetMapping("/file")
    public ResponseEntity<StreamingResponseBody> downloadFile(
            String path) throws UnsupportedEncodingException {

        String rootPath = getRootPath();
        if (rootPath == null) {
            throw new BackendException(rootPath, RespCode.ROOT_PATH_CONFIG_ERROR);
        }
        path = FilePathBuilder.ofPath()
                              .ct(rootPath)
                              .ct(path)
                              .build();

        File file = new File(path);

        // path is not valid
        if (!file.exists()) {
            throw new BackendException(path, RespCode.NO_FILE);
        }

        if (file.isDirectory()) {
            throw new BackendException(path, RespCode.DOWNLOAD_NOT_SUPPORT);
        }


        HttpHeaders headers = new HttpHeaders();
        headers.add("Access-Control-Expose-Headers", "Content-Disposition");
        headers.add("Content-Disposition",
                "attachment; filename=" + new String(file.getName()
                                                         .getBytes("UTF-8"), "iso-8859-1"));
        headers.add("Cache-Control", "no-cache, no-store, must-revalidate");
        headers.add("Pragma", "no-cache");
        headers.add("Expires", "0");
        headers.add("Content-Type", "application/octet-stream");
        headers.add(CONTENT_LENGTH, String.valueOf(file.length()));

        String finalPath = path;
        return ResponseEntity
                .ok()
                .headers(headers)
                .contentType(MediaType.parseMediaType("application/octet-stream"))
                .body(outputStream -> {
                    try (InputStream inputStream = Files.newInputStream(Paths.get(finalPath))) {
                        IOUtils.copyLarge(inputStream, outputStream);
                    }
                    catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                });
    }


    @PostMapping("/file")
    public ResponseEntity<JsonResult> uploadFile(
            @RequestPart("file-body") MultipartFile multipartFile,
            @RequestPart("file-info") @Validated UploadFileInfo uploadFileInfo
    ) throws Exception {

        // check if the size of upload file is valid
        long size = multipartFile.getSize();
        String maxSizeValue = configFile.getProp("max-upload-size");
        if(maxSizeValue != null && !maxSizeValue.equals("")) {
            long maxSize = Long.parseLong(maxSizeValue);
            maxSize *= 1024;
            maxSize *= 1024;
            if (size > maxSize) {
                throw new BackendException(null,
                        RespCode.UPLOAD_FILE_SIZE_EXCEED_UPPER_LIMIT);
            }
        } else {
            // config is not present, set default max size value, it can not
            // exceed configuration in application.yml
            long maxSize = 1024L * 1024 * 1024 * 2;
            if (size > maxSize) {
                throw new BackendException(null,
                        RespCode.UPLOAD_FILE_SIZE_EXCEED_UPPER_LIMIT);
            }
        }

        // file must have a name
        String filename = multipartFile.getOriginalFilename();
        if (filename == null) {
            throw new BackendException(null, RespCode.FILE_UPLOAD_ERROR);
        }

        // this is relative upload path
        String path = uploadFileInfo.getPath();

        // build the absolute path
        String rootPath = getRootPath();
        path = FilePathBuilder.ofPath()
                              .ct(rootPath)
                              .ct(path)
                              .build();


        File uploadDir = new File(path);
        if (uploadDir.isFile()) {
            throw new BackendException(path, RespCode.UPLOAD_DEST_ERROR);
        }

        Path uploadPath = Paths.get(path);

        // make upload dir if not exists
        if (!Files.exists(uploadPath)) {
            try {
                Files.createDirectories(uploadPath);
            }
            catch (java.nio.file.AccessDeniedException accessDeniedException) {
                throw new BackendException(path, RespCode.FILE_ACCESS_DENIED);
            }
        }

        Path uploadFilePath = uploadPath.resolve(filename);
        // check if file exists already
        boolean exists = Files.exists(uploadFilePath);
        if (exists) {
            throw new FileAlreadyExistsException(uploadFilePath.toString());
        }

        try {
            // copy file from multipart
            Files.createFile(uploadFilePath);
            Files.copy(
                    multipartFile.getInputStream(),
                    uploadFilePath,
                    StandardCopyOption.REPLACE_EXISTING);
        }
        catch (java.nio.file.AccessDeniedException accessDeniedException) {
            throw new BackendException(path, RespCode.FILE_ACCESS_DENIED);
        }


        return ResponseEntityUtils.ok(JsonResult.ok());

    }
}

File: src/main/java/fit/wenchao/http_file_server/rest/FileController.kt
package fit.wenchao.http_file_server.rest

import fit.wenchao.http_file_server.ConfigFile
import fit.wenchao.http_file_server.constants.API_PREFIX
import fit.wenchao.http_file_server.constants.DEFAULT_MAX_UPLOAD_FILE_SIZE
import fit.wenchao.http_file_server.constants.ONE_MB

import fit.wenchao.http_file_server.exception.BackendException
import fit.wenchao.http_file_server.exception.JsonResult
import fit.wenchao.http_file_server.exception.RespCode
import fit.wenchao.http_file_server.model.vo.FileInfo
import fit.wenchao.http_file_server.model.vo.UploadFileInfo
import fit.wenchao.http_file_server.rest.fileFilters.*
import fit.wenchao.http_file_server.service.AuthLogin
import fit.wenchao.http_file_server.service.FileService
import fit.wenchao.http_file_server.service.ThreadAuthContext
import fit.wenchao.http_file_server.service.UserAccessDirectoryService
import fit.wenchao.http_file_server.utils.FilePathBuilder
import fit.wenchao.http_file_server.utils.ResponseEntityUtils
import fit.wenchao.http_file_server.utils.iflet
import org.apache.commons.io.IOUtils
import org.springframework.context.ApplicationContext
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.nio.file.*
import javax.validation.Valid
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull

class FileListFilterOptsVO {

    @Valid
    var queryFilesOptions: MutableList<QueryFilesOptVO>? = mutableListOf()

    enum class FilterType {
        /**
         * value is a single entry, like: desc
         */
        SINGLE,

        /**
         * value is a list, like: 1, 2, 3
         */
        LIST
    }

    fun resolveFileListFilterOptions(): MutableList<QueryFilesOption> {
        var result = mutableListOf<QueryFilesOption>()
        queryFilesOptions?.forEach { filterConditionVO ->

            val key: String? = filterConditionVO.key
            val value: String? = filterConditionVO.value

            key ?: return@forEach
            value ?: return@forEach

            val type: String? = filterConditionVO.type
            // default type is SINGLE, Only support SINGLE now
            val confirmedType = type ?: FilterType.SINGLE.name

            if (confirmedType == FilterType.SINGLE.name) {
                val filterCondition = QueryFilesOption()
                filterCondition.key = key
                filterCondition.value = value
                filterCondition.type = confirmedType
                result.add(filterCondition)
            }

            // if (confirmedType == FilterType.LIST.name) {
            //
            // }

            // unknown filter type
            return@forEach
        }

        // we sort them in a preset way
        result = sortFileFilterOptions(result)

        return result
    }
}

data class QueryFilesOptVO(
    var type: String?,
    @NotNull
    var key: String?,
    @NotNull
    var value: String?,
) {

}

@Validated
@RestController
@RequestMapping(API_PREFIX)
class FileController(
    var appCtx: ApplicationContext,
    var configFile: ConfigFile,
    var fileService: FileService,
    var userAccessDirectoryService: UserAccessDirectoryService,
    var threadAuthContext: ThreadAuthContext
) {

    fun getRootPath(): String? {
        val entityId = threadAuthContext.getEntity()!!.getPrincipal().value()
        var root = userAccessDirectoryService.getUserAccessDirectory(entityId)
        return root
    }
    @AuthLogin
    @GetMapping("/file-list")
    fun getFileList(@RequestParam optsMap: MutableMap<String, String>, @NotBlank path: String): ResponseEntity<JsonResult> {

        // we get the filter options first
        var filterOptList = resolveFileListFilterOptions(optsMap)

        var filePath = path
        val rootPath = getRootPath()
        if (rootPath == "" || rootPath == null) {
            throw BackendException(rootPath, RespCode.ROOT_PATH_CONFIG_ERROR)
        }
        filePath = FilePathBuilder.ofPath()
            .ct(rootPath)
            .ct(filePath)
            .build()

        var fileInfos: MutableList<FileInfo>
        try {
            fileInfos = fileService.listFiles(filePath)
        }catch ( e: Exception){
            throw BackendException(null as Any?, rootPath, RespCode.ROOT_PATH_CONFIG_ERROR.getCode())
        }

        // we apply them to file list
        filterOptList.forEach { filterCondition ->
            val filter: Filter? = getFilter(appCtx, filterCondition)
            filter?.let {
                fileInfos = it.processFileList(fileInfos, filterCondition.value)
            }
        }

        return ResponseEntityUtils.ok(JsonResult.ok(fileInfos))


    }

    @AuthLogin
    @GetMapping("/file")
    fun downloadFile(
        @NotNull @NotBlank path: String
    ): ResponseEntity<StreamingResponseBody?>? {

        var filePath = path
        val rootPath = getRootPath()
        if (rootPath == "" || rootPath == null) {
            throw BackendException(rootPath, RespCode.ROOT_PATH_CONFIG_ERROR)
        }

        filePath = rootPath + filePath
//        filePath = FilePathBuilder.ofPath()
//            .ct(rootPath)
//            .ct(filePath)
//            .build()

        val file = File(filePath)

        // path is not valid
        if (!file.exists()) {
            throw BackendException(filePath, RespCode.NO_FILE)
        }
        if (file.isDirectory) {
            throw BackendException(filePath, RespCode.DOWNLOAD_NOT_SUPPORT)
        }
        val headers = HttpHeaders()
        headers.add("Access-Control-Expose-Headers", "Content-Disposition")
        headers.add(
            "Content-Disposition",
            "attachment; filename=" + String(
                file.name
                    .toByteArray(charset("UTF-8")), Charsets.ISO_8859_1
            )
        )
        headers.add("Cache-Control", "no-cache, no-store, must-revalidate")
        headers.add("Pragma", "no-cache")
        headers.add("Expires", "0")
        headers.add("Content-Type", "application/octet-stream")
        headers.add(HttpHeaders.CONTENT_LENGTH, file.length().toString())
        val finalPath = filePath
        return ResponseEntity
            .ok()
            .headers(headers)
            .contentType(MediaType.parseMediaType("application/octet-stream"))
            .body(StreamingResponseBody { outputStream: OutputStream? ->
                try {
                    Files.newInputStream(Paths.get(finalPath)).use { inputStream ->
                        IOUtils.copyLarge(
                            inputStream,
                            outputStream
                        )
                    }
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            })
    }

    @AuthLogin
    @PostMapping("/file")
    fun uploadFile(
        @RequestPart("file-body") multipartFile: MultipartFile,
        @RequestPart("file-info") @Validated uploadFileInfo: UploadFileInfo
    ): ResponseEntity<JsonResult> {

        // check if the size of upload file is valid
        val size = multipartFile.size
        val maxSizeValue: String? = configFile.getProp("max-upload-size")


        var maxSize: Long = 0L

        maxSizeValue.iflet {

            if (it == "") {
                maxSize = DEFAULT_MAX_UPLOAD_FILE_SIZE
            }

            maxSize = it.toLong()
            maxSize *= ONE_MB
        }.elselet {
            // config is not present, set default max size value, it can not
            // exceed configuration in application.yml
            maxSize = DEFAULT_MAX_UPLOAD_FILE_SIZE
        }

        if (size > maxSize) {
            throw BackendException(
                null,
                RespCode.UPLOAD_FILE_SIZE_EXCEED_UPPER_LIMIT
            )
        }

        // file must have a name
        val filename = multipartFile.originalFilename ?: throw BackendException(null, RespCode.FILE_UPLOAD_ERROR)

        // this is relative upload path
        var path = uploadFileInfo.path

        // build the absolute path
        val rootPath = getRootPath()
        if (rootPath == "" || rootPath == null) {
            throw BackendException(rootPath, RespCode.ROOT_PATH_CONFIG_ERROR)
        }
        path = FilePathBuilder.ofPath()
            .ct(rootPath)
            .ct(path)
            .build()
        val uploadDir = File(path)
        if (uploadDir.isFile) {
            throw BackendException(path, RespCode.UPLOAD_DEST_ERROR)
        }
        val uploadPath = Paths.get(path)

        // make upload dir if not exists
        if (!Files.exists(uploadPath)) {
            try {
                Files.createDirectories(uploadPath)
            } catch (accessDeniedException: AccessDeniedException) {
                throw BackendException(path, RespCode.FILE_ACCESS_DENIED)
            }
        }
        val uploadFilePath: Path = uploadPath.resolve(filename)
        // check if file exists already
        val exists = Files.exists(uploadFilePath)
        if (exists) {
            throw BackendException(null ,"File Exists", "FILE_EXISTS");
        }
        try {
            // copy file from multipart
            Files.createFile(uploadFilePath)
            Files.copy(
                multipartFile.inputStream,
                uploadFilePath,
                StandardCopyOption.REPLACE_EXISTING
            )
        } catch (accessDeniedException: AccessDeniedException) {
            throw BackendException(path, RespCode.FILE_ACCESS_DENIED)
        }
        return ResponseEntityUtils.ok(JsonResult.ok())
    }
}

fun main() {

}
